kubernetes/manifests/sync.go

```diff
@@ -0,0 +1,244 @@
+			deployment, err := clientset.AppsV1().Deployments(obj.GetNamespace()).Get(ctx, obj.GetName(), metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			if deployment.Status.ReadyReplicas != deployment.Status.Replicas || deployment.Status.UpdatedReplicas != deployment.Status.Replicas {
+				return retry.ExpectedErrorf("deployment %s ready replicas %d != replicas %d", deployment.Name, deployment.Status.ReadyReplicas, deployment.Status.Replicas)
+			}
+
+			logFunc(" > updated %s", deployment.GetName())
+
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func updateManifest(
+	ctx context.Context,
+	mapper *restmapper.DeferredDiscoveryRESTMapper,
+	k8sClient dynamic.Interface,
+	obj Manifest,
+	dryRun bool,
+) (
+	resp Manifest,
+	diff string,
+	skipped bool,
+	err error,
+) {
+	mapping, err := mapper.RESTMapping(obj.GroupVersionKind().GroupKind(), obj.GroupVersionKind().Version)
+	if err != nil {
+		err = fmt.Errorf("error creating mapping for object %s: %w", obj.GetName(), err)
+
+		return nil, "", false, err
+	}
+
+	var dr dynamic.ResourceInterface
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		// namespaced resources should specify the namespace
+		dr = k8sClient.Resource(mapping.Resource).Namespace(obj.GetNamespace())
+	} else {
+		// for cluster-wide resources
+		dr = k8sClient.Resource(mapping.Resource)
+	}
+
+	exists := true
+
+	diff, err = getResourceDiff(ctx, dr, obj)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, "", false, err
+		}
+
+		exists = false
+		diff = "resource is going to be created"
+	}
+
+	switch {
+	case dryRun:
+		return nil, diff, exists, nil
+	case !exists:
+		resp, err = dr.Create(ctx, obj, metav1.CreateOptions{})
+	case diff != "":
+		resp, err = dr.Update(ctx, obj, metav1.UpdateOptions{})
+	default:
+		skipped = true
+	}
+
+	return resp, diff, skipped, err
+}
+
+func getResourceDiff(ctx context.Context, dr dynamic.ResourceInterface, obj Manifest) (string, error) {
+	current, err := dr.Get(ctx, obj.GetName(), metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	obj.SetResourceVersion(current.GetResourceVersion())
+
+	resp, err := dr.Update(ctx, obj, metav1.UpdateOptions{
+		DryRun: []string{"All"},
+	})
+	if err != nil {
+		return "", err
+	}
+
+	ignoreKey := func(key ...string) {
+		unstructured.RemoveNestedField(current.Object, key...)
+		unstructured.RemoveNestedField(resp.Object, key...)
+	}
+
+	// drop fields which are not relevant and updated by Kubernetes
+	ignoreKey("metadata", "uid")
+	ignoreKey("metadata", "resourceVersion")
+	ignoreKey("metadata", "generation")
+	ignoreKey("metadata", "creationTimestamp")
+	ignoreKey("metadata", "deletionTimestamp")
+	ignoreKey("metadata", "deletionGracePeriodSeconds")
+	ignoreKey("metadata", "managedFields")
+	ignoreKey("metadata", "finalizers")
+	ignoreKey("metadata", "selfLink")
+	ignoreKey("metadata", "ownerReferences")
+	ignoreKey("metadata", "annotations")
```

Comment:
this is a user provided field, right?

Reply:
yep, but Kubernetes puts stuff there for Deployments... and it makes us remove that annotation. probably should go deeper here, like ignore only some annotations?

Reply:
working on a more specific version here....

Reply:
now it should be better, it ignores Kubernetes annotations, but allows to update user annotations.